sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+extern crate gstreamer_sdp_sys;
+extern crate gstreamer_webrtc;
+extern crate rand;
+extern crate ws;
+#[macro_use]
+extern crate serde_json;
+
+use gst::prelude::*;
+use gst::{BinExt, ElementExt};
+use rand::Rng;
+use std::sync::{Arc, Mutex};
+
+#[derive(PartialEq, PartialOrd, Eq, Debug)]
+enum AppState {
+    // AppStateUnknown = 0,
+    AppStateErr = 1,
+    ServerConnecting = 1000,
+    ServerConnectionError,
+    ServerConnected,
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallError,
+}
+#[derive(Debug)]
+enum Error {
+    GlibError(glib::Error),
+    GlibBoolError(glib::BoolError),
+    GstStateChangeError(gst::StateChangeError),
+}
+
+impl From<glib::Error> for Error {
+    fn from(error: glib::Error) -> Self {
+        Error::GlibError(error)
+    }
+}
+
+impl From<glib::BoolError> for Error {
+    fn from(error: glib::BoolError) -> Self {
+        Error::GlibBoolError(error)
+    }
+}
+
+impl From<gst::StateChangeError> for Error {
+    fn from(error: gst::StateChangeError) -> Self {
+        Error::GstStateChangeError(error)
+    }
+}
+
+const STUN_SERVER: &'static str = "stun-server=stun://stun.l.google.com:19302 ";
+const RTP_CAPS_OPUS: &'static str = "application/x-rtp,media=audio,encoding-name=OPUS,payload=";
+const RTP_CAPS_VP8: &'static str = "application/x-rtp,media=video,encoding-name=VP8,payload=";
+
+fn check_plugins() -> bool {
+    let needed = vec![
+        "opus",
+        "vpx",
+        "nice",
+        "webrtc",
+        "dtls",
+        "srtp",
+        "rtpmanager",
+        "videotestsrc",
+        "audiotestsrc",
+    ];
+    let registry = gst::Registry::get();
+    let mut ret = true;
+    for plugin_name in needed {
+        let plugin = registry.find_plugin(&plugin_name.to_string());
+        if plugin.is_none() {
+            println!("Required gstreamer plugin '{}' not found", plugin_name);
+            ret = false;
+        }
+    }
+    return ret;
+}
+
+fn setup_call(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::PeerConnecting;
+    println!(
+        "Setting up signalling server call with {}",
+        app_control.peer_id
+    );
+    app_control
+        .ws_sender
+        .send(format!("SESSION {}", app_control.peer_id))
+        .unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::ServerRegistering;
+    let our_id = rand::thread_rng().gen_range(10, 10_000);
+    println!("Registering id {} with server", our_id);
+    app_control
+        .ws_sender
+        .send(format!("HELLO {}", our_id))
+        .unwrap();
```

Comment:
Is the websocket API synchronous? Might make sense to do this asynchronously to not block any further processing

Reply:
Looks like it's synchronous. I'm unfamiliar with how to simply make a call async in rust without throwing it in a thread. Is there something in glib that could be leveraged? Or it seems the futures crate is used in gst-rs.

Reply:
Don't worry about that for now and just add a FIXME comment to that effect.

You can use glib for async things if the websocket implementation is integrated with the glib async infrastructure (which it isn't). The `ws` crate looks completely synchronous, I think what you need to do actually is to run `ws::connect()` on another thread and then communicate with that, and have another thread that does the GStreamer bus handling (and maybe generally uses the glib main loop).

Note this part on the `ws::connect()` function: `This function blocks until the event loop finishes running.`

Currently you would block the main thread when calling that function until all is done. There must be a way to somehow communicate from another thread with the socket or the handler so that you can integrate this with other application parts. This needs some changes and work, and probably also solves automatically your question about the bus handling. The answer is that you need to somehow have one event loop for the websocket and another for the bus (and potential other GLib things), or you have to integrate them with each other somehow. You need to check what mechanisms the `ws` crate provides for that.